Docs: add section on basic performance benchmark (#5635)
aiidateam/aiida-core: docs/source/howto/installation.rst

        and try a simple AiiDA query with the new database.
        If everything went fine, you can delete the old database location.

> It is not trivial to give a baseline of performance for AiiDA, as it depends a lot on the machine on which it is running, the installation, and what is being run.

Reviewer: ```suggestion
The performance for AiiDA depends both on the hardware it is running on and the processes being run.
```

Author: Wonder whether we should give more details here (or at least put it somewhere else). For example what is also important is how/where the services are running. If RabbitMQ and PostgreSQL are running on a different server, that is clearly going to incur a penalty versus them running on the localhost. Same goes for how the type of the filesystem and how it is mounted.

Author: That makes a lot of sense. I think I will do that and already add a section in Topics. Will probably be really minimal for now, but at least we'll have a placeholder that we can add onto in the future

> To still give some sense of performance, we provide this :download:`simple benchmarking script <include/scripts/performance_benchmark_base.py>` :fa:`download`.

Reviewer: ```suggestion
In order to get a rough idea, you can run :download:`simple benchmarking script <include/scripts/performance_benchmark_base.py>` :fa:`download`.
```

Download the script to the machine where AiiDA is running and make it executable.
Make sure the daemon is running (``verdi daemon start``) and then execute the script:

.. code:: console

    ./performance_benchmark_base.py -n 100

This will launch 100 ``ArithmeticAddCalculation`` jobs on the localhost and record the time until completion.

> Below is the output of a run using AiiDA v1.6.9 with a single daemon worker on a machine with a AMD Ryzen 5 3600 6-Core processor, with all services (RabbitMQ and PostgreSQL running on the same machine):

Reviewer: ```suggestion
Below is the output for running 100 ``ArithmeticAddCalculation``s with a single daemon worker on one core of an AMD Ryzen 5 3600 6-Core processor, using AiiDA v1.6.9, and RabbitMQ and PostgreSQL running on the same machine:
```

.. code:: console

    sph@citadel:~/$ ./performance_benchmark_base.py -n 100
        Success: Created and configured temporary `Computer` benchmark-5fa8c67f for localhost.
        Success: Created temporary `Code` bash for localhost.
        Submitting 100 calculations.  [####################################]  100%
        Submission completed in 9.36 seconds.
        Waiting for calculations to complete  [####################################]  100%
        Success: All calculations finished successfully.
        Elapsed time: 46.55 seconds.
        Cleaning up...
        Success: Deleted all calculations.
        Success: Deleted the created code bash@benchmark-5fa8c67f.
        Success: Deleted the created computer benchmark-5fa8c67f.
        Performance 2.15 processes / s

> As you can see, it took 9.42 seconds to submit the 100 calculations.
> The entire script took 46.55 seconds to submit and successfully complete all 100 calculations.
> When running the same script and setup with a different amount of daemon workers, we got the following numbers:

Reviewer: ```suggestionI
It took 9.42 seconds to submit the 100 calculations.
Overall, the entire script took 46.55 seconds to submit and successfully complete all 100 calculations, corresponding to a rate of roughly 2 calculations per second.

This rate can be sped up by adding more daemon workers:
```

> .. table::

Reviewer: This is a useful table.

If you want to see linear scaling, I guess you will need to remove the submission time.
I.e. stop the daemon, then submit the calculations, then start the daemon and measure how long the calculations take to complete (I assume this time will be roughly antiproportional to the number of daemons).

This should be straightforward with the python api you're adding for aiida v2 (but here it's indeed useful to have a v1-compatible script).

Author: Indeed, as you say, just removing the submission time is not a fair presentation either, because as soon as the first one is submitted, the daemon will start running those. Maybe I can include a note for this explaining the processes per second stat.

    :widths: auto

    ========== ======================= ==========================
    # Workers  Total elapsed time (s)  Performance (processes/s)
    ========== ======================= ==========================
    1          46.55                   2.15
    2          27.83                   3.59
    4          16.43                   6.09
    ========== ======================= ==========================

> Although there is no perfect linear scaling, an increase in the number of daemon workers clearly increases the throughput significantly.

Reviewer: ```suggestion
As expected, the process rate increases with the number of daemon workers.
The scaling is not linear, since the ``ArithmeticAddCalculation`` is very quick, and submission time is not influenced by the number of workers.
```
